File: Graph/Find_Eventual_Safe_States.py
#Problem Link :- https://leetcode.com/problems/find-eventual-safe-states/

"""In a directed graph, we start at some node and every turn, walk along a directed edge of 
the graph.  If we reach a node that is terminal (that is, it has no outgoing directed edges), 
we stop.Now, say our starting node is eventually safe if and only if we must eventually walk 
to a terminal node.  More specifically, there exists a natural number K so that for any 
choice of where to walk, we must have stopped at a terminal node in less than K steps.
Which nodes are eventually safe?  Return them as an array in sorted order"""

class Solution(object):
    def eventualSafeNodes(self, graph):
        ans = []
        n = len(graph)
        cycle = [False for i in range(n)]
        safe = [False for i in range(n)]
        visited = [False for i in range(n)]
        def helper(cur,graph,visited):
            if safe[cur] == True:
                return True
            if cycle[cur] == True:
                return False
            if visited[cur] == True:
                cycle[cur] = True
                return False
            visited[cur] = True
            for x in graph[cur]:
                if helper(x,graph,visited) == False:
                    cycle[x] = True
                    return False
            safe[cur] = True
            return True
        for i in range(len(graph)):
            if helper(i,graph,visited):
                ans.append(i)
        return ans

File: Graph/FindEventualSafeStates.cpp
//Problem Link:- https://leetcode.com/problems/find-eventual-safe-states/

#include<bits/stdc++.h>
using namespace std;

class Solution {
    private:
    vector<int> graph[10005];
public:
    vector<int> eventualSafeNodes(vector<vector<int>>& nums) {
        int nodeCount = nums.size();
        vector<int> indegree(nodeCount,0);
        for(int i=0;i<nodeCount;i++){
            vector<int> nie = nums[i];
            for(int j=0;j<nie.size();j++){
                graph[nie[j]].push_back(i);
                indegree[i]++;
            }
        }
        queue<int> pq;
        for(int i=0;i<nodeCount;i++){
            if (indegree[i] == 0){
                pq.push(i);
            }
        }
        vector<int> result;
        while (!pq.empty()){
            int node = pq.front();
            pq.pop();
            result.push_back(node);
            for(int child : graph[node]){
                indegree[child]--;
                if (indegree[child] == 0){
                    pq.push(child);
                }
            }
        }
        sort(result.begin(),result.end());
        return result;
    }
};